docs/src/adr/004adr-annotations.md

```diff
@@ -0,0 +1,136 @@
+# ADR-004: Syntax for Java-like annotations in comments
+
+**author:** Igor Konnov
+
+This ADR documents our decision on using Java-like annotations in comments.
+Our main motivation to have annotations is to simplify type annotations, as
+presented in [ADR-002][]. Hence, in the following text, we are using
+examples for type annotations. However, the annotations framework is not
+restricted to types. Similar to Java and Scala, we can use annotations
+to decorate operators with hints, which may aid the model checker.
+
+## 1. What can be annotated
+
+Annotations should be written in comments that are written in front of a
+declaration. The following declarations are supported:
+ 
+ 1. Constant declarations, e.g., `CONSTANT N`.
+ 1. Variable declarations, e.g., `VARIABLE x`.
+ 1. Top-level operator declarations, e.g., `Foo(x) == e`.
+ 1. LET-IN operator declarations, e.g., `Foo(x) == LET Inner(y) == e IN f`.
+ 1. Recursive functions, e.g., `foo[i \in Int] == e`.
+ 1. Recursive operators, that are declared either at the top-level or by using
+    LET-IN.
+
+For an example, see Section 3.
+
+
+## 2. Annotations syntax
+
+An annotation is a string that follows the grammar (question mark denotes
+optional rules):
+
+```
+Annotation  ::= '@' javaIdentifier ( '(' ArgList? ')' )?
+ArgList     ::= (Arg) ( ',' Arg )*
+Arg         ::= (string | integer | boolean)
+string      ::= '"' <char sequence> '"'
+integer     ::= '-'? [0-9]+
+boolean     ::= ('false' | 'true')
+```
+
+Java Language Specification defines how a [Java identifier] looks like.
+The sequence `<char sequence>` is a sequence admitted by [JavaTokenParsers]:
+  
+  - Any character except double quotes, control characters or backslash `\`
+  - A backslash followed by another backslash, a single or double quote,
+    or one of the letters `b`, `f`, `n`, `r` or `t`
+  - `\` followed by u followed by four hexadecimal digits
+  
+**Examples.** The following strings are examples of syntactically correct
+annotations:
+
+ - `@tailrec`
+ - `@type("(Int, Int) => Int")`
```

Comment:
since we're making up our own syntax, can't we do away with these ugly types as strings thing?

Reply:
I agree that it would be good to have a nicer syntax. But that seems to be hard, as SANY simply merges all strings in a comment together by removing `\n`.

Reply:
The reason why I'd like to use java annotations is that we can use them for other hints, e.g., `@pure` to indicate pure functions.

Reply:
Too bad :,(

Reply:
I know. If people start using our annotations, we can talk about extending the TLA+ parser :-)